File: csharp/common/Certificates/UniformCertificateStore.cs
/* 
 Copyright (c) 2010, Direct Project
 All rights reserved.

 Authors:
    Umesh Madan     umeshma@microsoft.com
  
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
Neither the name of The Direct Project (directproject.org) nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
*/
using System;
using System.Security.Cryptography.X509Certificates;

namespace Health.Direct.Common.Certificates
{
    /// <summary>
    /// Represents a certificate store where any address or domain uses the same collection of certificates
    /// Typically used at an organization level: where the same certificates apply to the entire organization.
    /// </summary>b
    public class UniformCertificateResolver : CertificateResolver
    {
        X509Certificate2Collection m_certs;
        
        /// <summary>
        /// Initializes an instance with the single certificate for the organization.
        /// </summary>
        /// <param name="cert">The certificate to use for the organization.</param>
        public UniformCertificateResolver(X509Certificate2 cert)
            : this(new X509Certificate2Collection(cert))
        {
        }

        /// <summary>
        /// Initializes an instance with a collection of certificates that may be used
        /// for the organization.
        /// </summary>
        /// <param name="certs">The certificates to use for the organization.</param>
        public UniformCertificateResolver(X509Certificate2Collection certs)
        {
            this.Certificates = certs;
        }

        /// <summary>
        /// Initializes an instance with a store of certificates that may be used
        /// for the organization.
        /// </summary>
        /// <param name="certs">The certificates to use for the organization.</param>
        public UniformCertificateResolver(IX509CertificateStore certs)
            : this(certs.GetAllCertificates())
        {
        }
        
        /// <summary>
        /// Gets and sets the certificates that may be used for the organization.
        /// </summary>
        public X509Certificate2Collection Certificates
        {
            get
            {
                return m_certs;
            }
            set
            {
                if (value == null)
                {
                    throw new ArgumentNullException("value");
                }

                m_certs = value;
            }
        }

        /// <summary>
        /// Actually resolves a certificate for the given name. Override to customize. 
        /// </summary>
        /// <param name="name">Return certificates for this name</param>
        /// <returns>
        /// The <see cref="X509Certificate2Collection"/> of certificates for the requested name.
        /// </returns>
        protected override X509Certificate2Collection Resolve(string name)
        {
            return m_certs;            
        }
    }
}

File: java/tags/agent-1.1.3/src/main/java/org/nhindirect/stagent/cert/impl/UniformCertificateStore.java
/* 
Copyright (c) 2010, NHIN Direct Project
All rights reserved.

Authors:
   Umesh Madan     umeshma@microsoft.com
   Greg Meyer      gm2552@cerner.com
 
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer 
in the documentation and/or other materials provided with the distribution.  Neither the name of the The NHIN Direct Project (nhindirect.org). 
nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE 
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
THE POSSIBILITY OF SUCH DAMAGE.
*/

package org.nhindirect.stagent.cert.impl;

import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;

import javax.mail.internet.InternetAddress;

import org.nhindirect.stagent.cert.CertificateResolver;
import org.nhindirect.stagent.cert.X509Store;
import org.nhindirect.stagent.cert.impl.annotation.UniformCertStoreCerts;

import com.google.inject.Inject;

/**
 * Certificate store when an entire organization is represented by a single certificate.  This should not be used if a single agent instance is
 * used to manage multiple organizations/domains.
 * @author Greg Meyer
 *
 */
public class UniformCertificateStore implements CertificateResolver
{
	private Collection<X509Certificate> certs;
	
    public UniformCertificateStore(X509Certificate cert)
    {
    	certs = new ArrayList<X509Certificate>();
    	certs.add(cert);
    }
    
    @Inject
    public UniformCertificateStore(@UniformCertStoreCerts Collection<X509Certificate> certs)
    {
    	setCertificates(certs);
    }    
    
    public UniformCertificateStore(X509Store certs)
    {
        if (certs == null)
        {
            throw new IllegalArgumentException();
        }
        this.setCertificates(certs.getAllCertificates());
    }    

    public void setCertificates(Collection<X509Certificate> certs)
	{
        if (certs == null || certs.size() == 0)
        {
            throw new IllegalArgumentException("Empty or null certificates are not allowed");
        }
        
        this.certs = new ArrayList<X509Certificate>(certs);
	}

    public Collection<X509Certificate> getCertificates(InternetAddress address)
	{
	    if (address == null)
	    {
	        throw new IllegalArgumentException();
	    }
	    
	    return Collections.unmodifiableCollection(certs);
	}
}
